Bsq swap: Implementation [3]

chimp1984 opened the pull request:
Based on https://github.com/bisq-network/bisq/pull/5770

Starts at 9c56202

This adds the whole BSQ swap implementation. I did not see much value in breaking it up into many commits. I think for review its best to check it out and do a full compare to the previous PR to see the changes.

All that work is derived from the work of @sqrrm  at https://github.com/bisq-network/bisq/pull/5729 

EDIT:
Merged and rebased  https://github.com/bisq-network/bisq/pull/5770

chimp1984 committed:
Move OfferPayload into package bisq.core.offer.bisq_v1

chimp1984 committed:
Move OfferPayload into package bisq.core.offer.bisq_v1

chimp1984 committed:
Move CreateOfferService into package bisq.core.offer.bisq_v1

chimp1984 committed:
Move CreateOfferService into package bisq.core.offer.bisq_v1

chimp1984 committed:
Move MarketPriceNotAvailableException into package bisq.core.offer.bisq_v1

chimp1984 committed:
Move MarketPriceNotAvailableException into package bisq.core.offer.bisq_v1

chimp1984 committed:
Move MutableOfferPayloadFields into package bisq.core.offer.bisq_v1

chimp1984 committed:
Move MutableOfferPayloadFields into package bisq.core.offer.bisq_v1

chimp1984 committed:
Move TriggerPriceService into package bisq.core.offer.bisq_v1

chimp1984 committed:
Move TriggerPriceService into package bisq.core.offer.bisq_v1

chimp1984 committed:
Move TakeOfferModel into package bisq.core.offer.bisq_v1

chimp1984 committed:
Move TakeOfferModel into package bisq.core.offer.bisq_v1

chimp1984 committed:
Move AvailabilityResult into package bisq.core.offer.availability

chimp1984 committed:
Move AvailabilityResult into package bisq.core.offer.availability

chimp1984 committed:
Move bisq.core.offer.messages package into package bisq.core.offer.availability

chimp1984 committed:
Move bisq.core.offer.messages package into package bisq.core.offer.availability

chimp1984 committed:
Move content of bisq.core.offer.placeoffer package into package bisq.core.offer.placeoffer.bisq_v1

chimp1984 committed:
Move content of bisq.core.offer.placeoffer package into package bisq.core.offer.placeoffer.bisq_v1

chimp1984 committed:
Rename OfferFilter to OfferFilterService

chimp1984 committed:
Rename OfferFilter to OfferFilterService

chimp1984 committed:
Move Contract, SellerAsMakerTrade, SellerAsTakerTrade, SellerTrade,
BuyerAsMakerTrade, BuyerAsTakerTrade, BuyerTrade
to bisq.core.trade.model.bisq_v1 package

chimp1984 committed:
Move Contract, SellerAsMakerTrade, SellerAsTakerTrade, SellerTrade,
BuyerAsMakerTrade, BuyerAsTakerTrade, BuyerTrade
to bisq.core.trade.model.bisq_v1 package

chimp1984 committed:
Make updateDepositTxFromWallet public

chimp1984 committed:
Make updateDepositTxFromWallet public

chimp1984 committed:
Move Trade to bisq.core.trade.model.bisq_v1 package

chimp1984 committed:
Move Trade to bisq.core.trade.model.bisq_v1 package

chimp1984 committed:
Move MakerTrade and TakerTrade to bisq.core.trade.model package

chimp1984 committed:
Move MakerTrade and TakerTrade to bisq.core.trade.model package

chimp1984 committed:
Move Tradable to bisq.core.trade.model package

chimp1984 committed:
Move Tradable to bisq.core.trade.model package

chimp1984 committed:
Move TradableList to bisq.core.trade.model package

chimp1984 committed:
Move TradableList to bisq.core.trade.model package

chimp1984 committed:
Make FluentProtocol.Event public

chimp1984 committed:
Make FluentProtocol.Event public

chimp1984 committed:
Make setup in FluentProtocol public

chimp1984 committed:
Make setup in FluentProtocol public

chimp1984 committed:
Move protocol classes to bisq.core.trade.protocol.bisq_v1 package

chimp1984 committed:
Move protocol classes to bisq.core.trade.protocol.bisq_v1 package

chimp1984 committed:
Move tasks package into bisq.core.trade.protocol.bisq_v1 package

chimp1984 committed:
Move tasks package into bisq.core.trade.protocol.bisq_v1 package

chimp1984 committed:
Move TradingPeer into bisq.core.trade.protocol.bisq_v1 package

chimp1984 committed:
Move TradingPeer into bisq.core.trade.protocol.bisq_v1 package

chimp1984 committed:
Move messages package into bisq.core.trade.protocol.bisq_v1 package
Move TradeMessage into bisq.core.trade.protocol package

chimp1984 committed:
Move messages package into bisq.core.trade.protocol.bisq_v1 package
Move TradeMessage into bisq.core.trade.protocol package

chimp1984 committed:
Move ClosedTradableManager, FailedTradesManager into bisq.core.trade.bisq_v1 package

chimp1984 committed:
Move ClosedTradableManager, FailedTradesManager into bisq.core.trade.bisq_v1 package

chimp1984 committed:
Rename CleanupMailboxMessages to CleanupMailboxMessagesService

chimp1984 committed:
Rename CleanupMailboxMessages to CleanupMailboxMessagesService

chimp1984 committed:
Move CleanupMailboxMessagesService, ClosedTradeUtil to package bisq.core.trade.bisq_v1

chimp1984 committed:
Move CleanupMailboxMessagesService, ClosedTradeUtil to package bisq.core.trade.bisq_v1

chimp1984 committed:
Move CleanupMailboxMessagesService, ClosedTradeUtil to package bisq.core.trade.bisq_v1

chimp1984 committed:
Move TransactionResultHandler, TradeResultHandler to package bisq.core.trade.bisq_v1

chimp1984 committed:
Move TransactionResultHandler, TradeResultHandler to package bisq.core.trade.bisq_v1

chimp1984 committed:
Move TransactionResultHandler, TradeResultHandler to package bisq.core.trade.bisq_v1

chimp1984 committed:
Move TradeDataValidation, DumpDelayedPayoutTx, TradeDataValidation, TradeUtil, TradeTxException to package bisq.core.trade.bisq_v1

chimp1984 committed:
Move TradeDataValidation, DumpDelayedPayoutTx, TradeDataValidation, TradeUtil, TradeTxException to package bisq.core.trade.bisq_v1

chimp1984 committed:
Move TradeDataValidation, DumpDelayedPayoutTx, TradeDataValidation, TradeUtil, TradeTxException to package bisq.core.trade.bisq_v1

chimp1984 committed:
Rename ProcessModelServiceProvider to Provider

chimp1984 committed:
Rename ProcessModelServiceProvider to Provider

chimp1984 committed:
Rename ProcessModelServiceProvider to Provider

chimp1984 committed:
Rename txID to txId

chimp1984 committed:
Rename txID to txId

chimp1984 committed:
Rename txID to txId

chimp1984 committed:
Move createoffer package inside bisq.desktop.main.offer.bisq_v1 package

chimp1984 committed:
Move createoffer package inside bisq.desktop.main.offer.bisq_v1 package

chimp1984 committed:
Move createoffer package inside bisq.desktop.main.offer.bisq_v1 package

chimp1984 committed:
Move takeoffer package inside bisq.desktop.main.offer.bisq_v1 package

chimp1984 committed:
Move takeoffer package inside bisq.desktop.main.offer.bisq_v1 package

chimp1984 committed:
Move takeoffer package inside bisq.desktop.main.offer.bisq_v1 package

chimp1984 committed:
Move MutableOffer view classes into bisq.desktop.main.offer.bisq_v1 package

chimp1984 committed:
Move MutableOffer view classes into bisq.desktop.main.offer.bisq_v1 package

chimp1984 committed:
Move MutableOffer view classes into bisq.desktop.main.offer.bisq_v1 package

chimp1984 committed:
Move OfferDataModel, OfferViewUtil into bisq.desktop.main.offer.bisq_v1 package

chimp1984 committed:
Move OfferDataModel, OfferViewUtil into bisq.desktop.main.offer.bisq_v1 package

chimp1984 committed:
Move OfferDataModel, OfferViewUtil into bisq.desktop.main.offer.bisq_v1 package

chimp1984 committed:
Rename FeeUtil to OfferViewModelUtil

chimp1984 committed:
Rename FeeUtil to OfferViewModelUtil

chimp1984 committed:
Rename FeeUtil to OfferViewModelUtil

chimp1984 committed:
Move OfferViewModelUtil into bisq.desktop.main.offer.bisq_v1 package

chimp1984 committed:
Move OfferViewModelUtil into bisq.desktop.main.offer.bisq_v1 package

chimp1984 committed:
Move OfferViewModelUtil into bisq.desktop.main.offer.bisq_v1 package

chimp1984 committed:
Cleanups, refromatting, add final qualifier

chimp1984 committed:
Cleanups, refromatting, add final qualifier

chimp1984 committed:
Cleanups, refromatting, add final qualifier

chimp1984 committed:
Move TradingPeer to package bisq.core.trade.protocol.bisq_v1.model

chimp1984 committed:
Move TradingPeer to package bisq.core.trade.protocol.bisq_v1.model

chimp1984 committed:
Move TradingPeer to package bisq.core.trade.protocol.bisq_v1.model

chimp1984 committed:
Make setPaymentStartedAckMessage and setDepositTxSentAckMessage public

chimp1984 committed:
Make setPaymentStartedAckMessage and setDepositTxSentAckMessage public

chimp1984 committed:
Make setPaymentStartedAckMessage and setDepositTxSentAckMessage public

chimp1984 committed:
Move ProcessModel to package bisq.core.trade.protocol.bisq_v1.model

chimp1984 committed:
Move ProcessModel to package bisq.core.trade.protocol.bisq_v1.model

chimp1984 committed:
Move ProcessModel to package bisq.core.trade.protocol.bisq_v1.model

chimp1984 committed:
Cleanups

chimp1984 committed:
Cleanups

chimp1984 committed:
Add linebreak in error popup
Remove delay for success popup

chimp1984 committed:
Fix layout/text issues

chimp1984 committed:
Increase visibility

chimp1984 committed:
Remove Inject annotation at abstract classes

chimp1984 committed:
Add delegate methods to offer
Remove unused methods

chimp1984 committed:
Move Offer.Direction one level up and rename to OfferDirection

chimp1984 committed:
Rename paymentMethod.isAsset to paymentMethod.isBlockchain

We will need later to distinguish between BSQ and other Blockchain assets

chimp1984 committed:
Add isFiat and isAltcoin methods.
Use isFiat instead of negation of isBlockchain.
Use isAltcoin if all non btc assets are targeted

chimp1984 committed:
Rename availableConfirmedBalance to availableBalance

It does not include only confirmed txs as we include
the unconfirmed own change outputs.

chimp1984 committed:
Only check for localhost btc node if mainnet

chimp1984 committed:
Make TxConfidenceListener and its onTransactionConfidenceChanged method abstract
Use lombok getter

chimp1984 committed:
Add scriptTypeId to RawTransactionInput

Add convenience methods

chimp1984 committed:
Add deprecated annotation, null checks and comment to getRawInputFromTransactionInput

chimp1984 committed:
Move test classes into bisq.desktop.main.offer.bisq_v1.createoffer

chimp1984 committed:
Merge branch 'bsq-swap-low-risk-refactorings' into wip-merge2

# Conflicts:
#	core/src/main/java/bisq/core/account/witness/AccountAgeWitnessService.java
#	core/src/main/java/bisq/core/api/CoreOffersService.java
#	core/src/main/java/bisq/core/btc/listeners/TxConfidenceListener.java
#	core/src/main/java/bisq/core/notifications/alerts/market/MarketAlerts.java
#	core/src/main/java/bisq/core/offer/Offer.java
#	core/src/main/java/bisq/core/offer/OfferForJson.java
#	core/src/main/java/bisq/core/offer/OfferUtil.java
#	core/src/main/java/bisq/core/offer/bisq_v1/TakeOfferModel.java
#	core/src/main/java/bisq/core/proto/ProtoDevUtil.java
#	core/src/main/java/bisq/core/trade/TradeManager.java
#	core/src/main/java/bisq/core/trade/statistics/TradeStatistics2.java
#	core/src/test/java/bisq/core/offer/OfferMaker.java
#	desktop/src/main/java/bisq/desktop/components/paymentmethods/PaymentMethodForm.java
#	desktop/src/main/java/bisq/desktop/main/PriceUtil.java
#	desktop/src/main/java/bisq/desktop/main/market/offerbook/OfferBookChartView.java
#	desktop/src/main/java/bisq/desktop/main/market/offerbook/OfferBookChartViewModel.java
#	desktop/src/main/java/bisq/desktop/main/market/spread/SpreadViewModel.java
#	desktop/src/main/java/bisq/desktop/main/offer/BuyOfferView.java
#	desktop/src/main/java/bisq/desktop/main/offer/OfferView.java
#	desktop/src/main/java/bisq/desktop/main/offer/SellOfferView.java
#	desktop/src/main/java/bisq/desktop/main/offer/bisq_v1/MutableOfferDataModel.java
#	desktop/src/main/java/bisq/desktop/main/offer/bisq_v1/MutableOfferView.java
#	desktop/src/main/java/bisq/desktop/main/offer/bisq_v1/MutableOfferViewModel.java
#	desktop/src/main/java/bisq/desktop/main/offer/bisq_v1/takeoffer/TakeOfferDataModel.java
#	desktop/src/main/java/bisq/desktop/main/offer/bisq_v1/takeoffer/TakeOfferView.java
#	desktop/src/main/java/bisq/desktop/main/offer/bisq_v1/takeoffer/TakeOfferViewModel.java
#	desktop/src/main/java/bisq/desktop/main/offer/offerbook/OfferBook.java
#	desktop/src/main/java/bisq/desktop/main/offer/offerbook/OfferBookView.java
#	desktop/src/main/java/bisq/desktop/main/offer/offerbook/OfferBookViewModel.java
#	desktop/src/main/java/bisq/desktop/main/overlays/windows/OfferDetailsWindow.java
#	desktop/src/main/java/bisq/desktop/main/portfolio/closedtrades/ClosedTradesDataModel.java
#	desktop/src/main/java/bisq/desktop/main/portfolio/editoffer/EditOfferDataModel.java
#	desktop/src/main/java/bisq/desktop/main/portfolio/failedtrades/FailedTradesDataModel.java
#	desktop/src/main/java/bisq/desktop/main/portfolio/openoffer/OpenOffersDataModel.java
#	desktop/src/main/java/bisq/desktop/main/portfolio/pendingtrades/PendingTradesDataModel.java
#	desktop/src/main/java/bisq/desktop/util/DisplayUtils.java
#	desktop/src/test/java/bisq/desktop/main/offer/bisq_v1/createoffer/CreateOfferDataModelTest.java
#	desktop/src/test/java/bisq/desktop/main/offer/bisq_v1/createoffer/CreateOfferViewModelTest.java
#	desktop/src/test/java/bisq/desktop/main/offer/offerbook/OfferBookListItemMaker.java
#	desktop/src/test/java/bisq/desktop/main/portfolio/editoffer/EditOfferDataModelTest.java
#	desktop/src/test/java/bisq/desktop/maker/OfferMaker.java

chimp1984 committed:
Fix merge issues

chimp1984 committed:
Add allowSpendMyOwnUnconfirmedTxOutputs flag to BsqCoinSelector
Extract variable for TxOutputKey

chimp1984 committed:
Add isMine method. Add error logs

chimp1984 committed:
Inline TransactionWitness
Add line breaks, cleanups

chimp1984 committed:
Add new methods

chimp1984 committed:
Add signTx and verifyNonDustTxo methods

chimp1984 committed:
Remove unused signTx method

chimp1984 committed:
Add verifiedBalance field.
Use WalletService.signTx and WalletService.verifyNonDustTxo inside signTx instance method

chimp1984 committed:
Rename signTx to signTxAndVerifyNoDustOutputs (no code change)

chimp1984 committed:
Change log level

chimp1984 committed:
Add comment, cleanups

chimp1984 committed:
Rename processModel.getTradingPeer to getTradePeer

chimp1984 committed:
Only print stack trace if dev mode

chimp1984 committed:
Change log level

chimp1984 committed:
Replace Value annotation with Getter and EqualsAndHashCode

chimp1984 committed:
Add methods

chimp1984 committed:
Simplify

chimp1984 committed:
Move getRandomOfferId method to OfferUtil
Add more methods to OfferUtil
Cleanups

chimp1984 committed:
Replace Value with Getter

chimp1984 committed:
Rename trade.getPhase to getTradePhase

chimp1984 committed:
Rename trade.getState to getTradeState

chimp1984 committed:
Cleanup

chimp1984 committed:
Extract methods, Cleanups

chimp1984 committed:
Extract methods, cleanups. Rename methods and vars

chimp1984 committed:
Add exclude fields to EqualsAndHashCode
Add constructor (will be used for bsq swaps)

Not 100% sure if that is valid.

chimp1984 committed:
Rename trade.getPhase to getTradePhase
Rename getTakeOfferDate to getDate

chimp1984 committed:
Extract to method in TradeUtil

chimp1984 committed:
Use early return. cleanup

chimp1984 committed:
Extract method. Rename vars

chimp1984 committed:
Cleanups, apply rename of getTradePhase

chimp1984 committed:
Add method. Dont return null in case of no amount but min fee

chimp1984 committed:
Add method. cleanups

chimp1984 committed:
Move OfferDirection from bisq.core.offer.bisq_v1 to bisq.core.offer

chimp1984 committed:
Remove handling of BundleOfEnvelopes in send method.
It is basically never executed and throttling would avoid it as well.
Cleanups, Rename

chimp1984 committed:
Cleanups

chimp1984 committed:
Add filterPredicate

chimp1984 committed:
Make onDeleteAccount more clear

chimp1984 committed:
Only add delay if not devmode

chimp1984 committed:
Reduce visibility

chimp1984 committed:
Add methods, move field up

chimp1984 committed:
Add method

chimp1984 committed:
Add style

chimp1984 committed:
Add linebreak (no code change)
Cleanup

chimp1984 committed:
Add isNotPublished method (will be combined later with bsq swap checks)
Cleanups

chimp1984 committed:
Add null checks, extract variables, cleanups

chimp1984 committed:
Move tests to package bisq.desktop.main.offer.bisq_v1.createoffer

chimp1984 committed:
Move tests to package bisq.desktop.main.offer.bisq_v1.createoffer

chimp1984 committed:
Merge branch 'chimp-bsq-swap' into wip-merge021.10

# Conflicts:
#	desktop/src/test/java/bisq/desktop/main/offer/bisq_v1/createoffer/CreateOfferDataModelTest.java

chimp1984 committed:
Dummy commit to enforce rebuild at CI

chimp1984 committed:
Merge branch 'chimp-bsq-swap' into bsq-swap-low-risk-refactorings

chimp1984 committed:
Add abstract classes and interfaces. Let ProcessModel and TradingPeer extend those.

chimp1984 committed:
Let Trade extend TradeModel

chimp1984 committed:
Generify TradeResultHandler

chimp1984 committed:
Use TradeModel and ProtocolModel in TradeTaskRunner, TradeTask, TradeProtocol and FluentProtocol

chimp1984 committed:
Use TradeModel

chimp1984 committed:
Impl. onAckMessage. Add fields for concrete types

chimp1984 committed:
Use TradeModel in trade domain

chimp1984 committed:
Let OfferPayload extend OfferPayloadBase

chimp1984 committed:
Use TradeModel instead of Trade where appropriate.
Extend methods with additional params.
Cleanups

chimp1984 committed:
Move protobuf OfferPayload.Direction to OfferDirection

chimp1984 committed:
Move protobuf OfferPayload.Direction to OfferDirection

chimp1984 committed:
Use bitcoinj 42bbae9 (with fix required for bsq swaps)

chimp1984 committed:
Add support for RemoveDataMessage in testCapability
Refactor testCapability

chimp1984 committed:
Cleanups

chimp1984 committed:
Remove todo

chimp1984 committed:
Add BSQ swap code

chimp1984 committed:
Apply codacy review suggestions

chimp1984 commented:
No idea why the gradle dependencyVerification fails. It is the same content as in the release branch.

chimp1984 commented:
Seems its a jitpack issue. Devs on keybase complained also about problems.

chimp1984 committed:
Move Utilities.objectToJson to JsonUtil.objectToJson
Move inner class AnnotationExclusionStrategy to JsonUtil
Next commit includes changes why that move to core is needed.

chimp1984 committed:
Add custom json serializer for OfferPayload

Order of fields is different and that would cause
a diff. json and diff. hash for the contract.

chimp1984 committed:
Update text

chimp1984 committed:
Remove missingFundsListener after successful take offer

chimp1984 committed:
Change log level

chimp1984 committed:
Improve BSQ swap offer details window

chimp1984 committed:
Update dev mode values for create offer

chimp1984 committed:
Workaround for weird JFXComboBox bug

The currency list would show up the list overlay. That gets triggered by unclear reasons.
The bindings on the action and cancel buttons are causing that behaviour (maybe causes some
render cycle which then causes the issue...). By requesting focus on our view directly after
the onTakeOffer updates the bindable properties we can get rid of that UI glitch.

chimp1984 committed:
Add default value if no filter is available

chimp1984 committed:
Add log

chimp1984 committed:
Reset address entry if BSQ swap offer is used for create offer

We create a new address entry when opening the create offer view.
But we close and replace that view with the BSQ swap create offer view if
BSQ swap is selected, therefore the address entry would be dangling and
later cleaned up at openOfferManager. To clean it up directly we call the
resetAddressEntriesForOpenOffer on btcWalletService.

chimp1984 committed:
Use checkNotNull and checkArgument for checking pre conditions
in signBsqSwapTransaction.
Remove unused code

chimp1984 committed:
Add requestPersistence call.
Extract variables, inline method

chimp1984 committed:
Improve naming, comments, extract variables

chimp1984 committed:
Improve naming, comments, extract variables, inline method

chimp1984 committed:
Remove redundant complete call

chimp1984 committed:
Improve naming, cleanups, use static imports

chimp1984 committed:
Fix linebreak

chimp1984 committed:
Combine nested if statements

chimp1984 committed:
Revert behaviour to old version in master to return null if amount is null.
This caused an issue in the UI so its more safe to leave old behaviour.

chimp1984 committed:
Cleanups

ghubstan commented on desktop/src/main/java/bisq/desktop/main/dao/DaoView.java:
```diff
@@ -145,8 +145,12 @@ public void initialize() {
+            if (preferences.showAgain(DaoPresentation.DAO_NEWS)) {
+                preferences.dontShowAgain(DaoPresentation.DAO_NEWS, true);
+                new Popup().headLine(Res.get("news.bsqSwap.title"))
+                        .information(Res.get("news.bsqSwap.description"))
```
TODO Write and publish [BSQ Swaps wiki.](https://bisq.wiki/BSQ_Swaps)

chimp1984 replied:
I pinged @m52go . I think he had already the content but I dont see a PR yet.

ghubstan commented on core/src/main/java/bisq/core/offer/placeoffer/bsqswap/PlaceBsqSwapOfferModel.java:
```diff
@@ -0,0 +1,46 @@
+ * along with Bisq. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package bisq.core.offer.placeoffer.bsqswap;
```
Some packages are named `bsq_swap`, but this one is `bsqswap`.  Could the `bsq_swap` pkgs be renamed `bsqswap` too?

chimp1984 replied:
Ops. Renamed in a new commit.

ghubstan commented on desktop/src/main/java/bisq/desktop/main/overlays/windows/BsqSwapOfferDetailsWindow.java:
```diff
@@ -0,0 +1,320 @@
+        String minus = " - ";
+        String plus = " + ";
+        String minerFeePostFix = Res.get("tradeDetailsWindow.txFee");
+        String tradeFeePostFix = Res.get("shared.tradeFee");
```
Is there no way to tell the bsq swap offer creator what the trade fee will be
![how-much-is-trade-fee](https://user-images.githubusercontent.com/36207203/139327043-ca7084ae-94d9-45d2-8773-1e6cdacf69c2.png)
?

ghubstan replied:
The taker can see the trade fee before taking the offer, but the maker still doesn't know yet.

ghubstan replied:
After taking the offer, the maker still can't find the trade fee, until he checks the swap tx in a block explorer, or exporting swap trade to CSV file.

chimp1984 replied:
When creating the offer the second step shows all the fees. I was a bit lazy to add it here as well and as the tx fee are calculated based on the inputs/change its not that trivial. But yes could be added.

chimp1984 committed:
Apply patch from ghubstan

Remove currency params. Bsq swaps have statically defined currencies (BSQ/BTC).

chimp1984 committed:
Rename package

chimp1984 committed:
Fix link to BSQ_Swaps wiki

chimp1984 committed:
Apply patch from @ghubstan

chimp1984 committed:
Merge branch 'chimp-bsq-swap-1' into bsq-swap-low-risk-refactorings-2

# Conflicts:
#	core/src/main/java/bisq/core/account/witness/AccountAgeWitnessService.java
#	core/src/main/java/bisq/core/api/CoreOffersService.java
#	core/src/main/java/bisq/core/btc/listeners/TxConfidenceListener.java
#	core/src/main/java/bisq/core/notifications/alerts/market/MarketAlerts.java
#	core/src/main/java/bisq/core/offer/OpenOfferManager.java
#	core/src/main/java/bisq/core/offer/bisq_v1/CreateOfferService.java
#	core/src/main/java/bisq/core/offer/bisq_v1/OfferPayload.java
#	core/src/main/java/bisq/core/offer/bisq_v1/TakeOfferModel.java
#	core/src/main/java/bisq/core/offer/bisq_v1/TriggerPriceService.java
#	core/src/main/java/bisq/core/proto/ProtoDevUtil.java
#	core/src/main/java/bisq/core/trade/TradeManager.java
#	core/src/main/java/bisq/core/trade/bisq_v1/TradeResultHandler.java
#	core/src/main/java/bisq/core/trade/bisq_v1/TradeUtil.java
#	core/src/main/java/bisq/core/trade/model/bisq_v1/Trade.java
#	core/src/main/java/bisq/core/trade/protocol/FluentProtocol.java
#	core/src/main/java/bisq/core/trade/protocol/Provider.java
#	core/src/main/java/bisq/core/trade/protocol/TradeProtocol.java
#	core/src/main/java/bisq/core/trade/protocol/TradeProtocolFactory.java
#	core/src/main/java/bisq/core/trade/protocol/TradeTaskRunner.java
#	core/src/main/java/bisq/core/trade/protocol/bisq_v1/DisputeProtocol.java
#	core/src/main/java/bisq/core/trade/protocol/bisq_v1/model/ProcessModel.java
#	core/src/main/java/bisq/core/trade/protocol/bisq_v1/tasks/ApplyFilter.java
#	core/src/main/java/bisq/core/trade/protocol/bisq_v1/tasks/TradeTask.java
#	core/src/main/java/bisq/core/trade/protocol/bisq_v1/tasks/maker/MakerCreateAndSignContract.java
#	core/src/main/java/bisq/core/trade/protocol/bisq_v1/tasks/taker/TakerVerifyAndSignContract.java
#	core/src/main/java/bisq/core/trade/statistics/TradeStatistics2.java
#	core/src/main/java/bisq/core/trade/statistics/TradeStatisticsManager.java
#	core/src/main/java/bisq/core/trade/txproof/xmr/XmrTxProofService.java
#	desktop/src/main/java/bisq/desktop/components/paymentmethods/PaymentMethodForm.java
#	desktop/src/main/java/bisq/desktop/main/PriceUtil.java
#	desktop/src/main/java/bisq/desktop/main/funds/transactions/TransactionAwareTradableFactory.java
#	desktop/src/main/java/bisq/desktop/main/funds/transactions/TransactionAwareTrade.java
#	desktop/src/main/java/bisq/desktop/main/market/offerbook/OfferBookChartView.java
#	desktop/src/main/java/bisq/desktop/main/market/offerbook/OfferBookChartViewModel.java
#	desktop/src/main/java/bisq/desktop/main/market/spread/SpreadViewModel.java
#	desktop/src/main/java/bisq/desktop/main/offer/BuyOfferView.java
#	desktop/src/main/java/bisq/desktop/main/offer/OfferView.java
#	desktop/src/main/java/bisq/desktop/main/offer/SellOfferView.java
#	desktop/src/main/java/bisq/desktop/main/offer/bisq_v1/MutableOfferDataModel.java
#	desktop/src/main/java/bisq/desktop/main/offer/bisq_v1/MutableOfferView.java
#	desktop/src/main/java/bisq/desktop/main/offer/bisq_v1/MutableOfferViewModel.java
#	desktop/src/main/java/bisq/desktop/main/offer/bisq_v1/takeoffer/TakeOfferDataModel.java
#	desktop/src/main/java/bisq/desktop/main/offer/bisq_v1/takeoffer/TakeOfferView.java
#	desktop/src/main/java/bisq/desktop/main/offer/bisq_v1/takeoffer/TakeOfferViewModel.java
#	desktop/src/main/java/bisq/desktop/main/offer/offerbook/OfferBook.java
#	desktop/src/main/java/bisq/desktop/main/offer/offerbook/OfferBookView.java
#	desktop/src/main/java/bisq/desktop/main/offer/offerbook/OfferBookViewModel.java
#	desktop/src/main/java/bisq/desktop/main/overlays/windows/OfferDetailsWindow.java
#	desktop/src/main/java/bisq/desktop/main/portfolio/closedtrades/ClosedTradesDataModel.java
#	desktop/src/main/java/bisq/desktop/main/portfolio/editoffer/EditOfferDataModel.java
#	desktop/src/main/java/bisq/desktop/main/portfolio/failedtrades/FailedTradesDataModel.java
#	desktop/src/main/java/bisq/desktop/main/portfolio/openoffer/OpenOffersDataModel.java
#	desktop/src/main/java/bisq/desktop/main/portfolio/openoffer/OpenOffersView.java
#	desktop/src/main/java/bisq/desktop/main/portfolio/pendingtrades/PendingTradesDataModel.java
#	desktop/src/main/java/bisq/desktop/util/DisplayUtils.java
#	desktop/src/test/java/bisq/desktop/main/offer/bisq_v1/createoffer/CreateOfferDataModelTest.java
#	desktop/src/test/java/bisq/desktop/main/offer/bisq_v1/createoffer/CreateOfferViewModelTest.java
#	desktop/src/test/java/bisq/desktop/main/offer/offerbook/OfferBookListItemMaker.java
#	desktop/src/test/java/bisq/desktop/main/portfolio/editoffer/EditOfferDataModelTest.java
#	desktop/src/test/java/bisq/desktop/maker/OfferMaker.java
#	monitor/src/main/java/bisq/monitor/metric/P2PMarketStats.java

chimp1984 committed:
Fix merge issues (nothing critical)

chimp1984 committed:
Apply patch from @ghubstan

chimp1984 committed:
Add missing class (merge mistake)

chimp1984 committed:
Merge branch 'bsq-swap-low-risk-refactorings-2' into bsq-swap-impl-3

ripcurlx commented:
I just tried both scenarios to Buy and Sell BTC for BSQ  successfully! 🚀  These are my comments for the UI parts of the trade process.

1) I think it would be good to mention in the popup or in wiki article linked in the `NEW` popup (https://bisq.wiki/BSQ_swaps), that a hidden BSQ Swap account is created: `BSQ Swap (BSQ, BSQ Swap)`. It wasn't a big hurdle when I tried to create a Buy BTC offer that I need to select the new BSQ Swap account, but I think it would be good to mention that this account is created automatically and that it is not visible in the account section as all other accounts.

![Bildschirmfoto 2021-11-03 um 10 16 23](https://user-images.githubusercontent.com/170962/140039555-86b4ee63-ecdc-4627-bf4b-ac2ce466983e.png)

2) When creating a Buy BTC (Sell BSQ) offer I recognized that after creating the offer the filter by currency combobox was selected and open (I have the same open issue in the buy BSQ branch). Is this a known problem?
![Bildschirmfoto 2021-11-03 um 10 21 26](https://user-images.githubusercontent.com/170962/140039815-7740dc15-a6a5-424d-ba2b-2caf25defdf8.png)

3) I'm not sure if it is necessary to change the label of the tab to `Create BSQ Swap offer` from `Create offer` when a BSQ swap account is selected, as it is also placed under `My open offers`. 
![Bildschirmfoto 2021-11-03 um 10 20 23](https://user-images.githubusercontent.com/170962/140040110-3918434c-aaaf-4b35-a912-55ea8abaf5cd.png)

4) When taking a BSQ swap offer, after taking the offer successfully I get following success message.

![Bildschirmfoto 2021-11-03 um 10 24 50](https://user-images.githubusercontent.com/170962/140040400-998a9a7d-00d3-4a34-8f11-d4777d5ad85b.png)

This leaves me with the impression that the trade is already completed.  I was expecting in that case that I'll find  the trade in my portfolio > history tab. But as mentioned in the overlay above there is a new BSQ SWAP TRADES tab (Should be in lower case to match other navigation references: `BSQ swap trades`).  So I though that this is because the transaction doesn't have already a block confirmation.

![Bildschirmfoto 2021-11-03 um 10 25 43](https://user-images.githubusercontent.com/170962/140040902-d86bb599-d39e-4187-8d28-78d0e36394b2.png)
![Bildschirmfoto 2021-11-03 um 10 26 37](https://user-images.githubusercontent.com/170962/140040916-ee65f0d9-216f-476b-9ebc-1ab584dd2c9c.png)
After sufficient confirmations the tab is still there. Is there a reason why we didn't call this tab `Open BSQ swap trades` and move it to history after the trade has at least one block confirmation? I personally would appreciate having all my trades in one history, or is there a technical reason for this?

cd2357 commented:
> No idea why the gradle dependencyVerification fails. It is the same content as in the release branch.

The latest build (re-triggered by @ripcurlx 20 mins ago) was successful. Should also work locally.

ripcurlx commented:
![Bildschirmfoto 2021-11-03 um 11 28 54](https://user-images.githubusercontent.com/170962/140044708-786744ec-9c57-4447-afe7-a15ee589dd01.png)
![Bildschirmfoto 2021-11-03 um 11 29 04](https://user-images.githubusercontent.com/170962/140044712-a449c372-b6c9-46e8-a310-2cad087308d7.png)

I also just recognized that all altcoin/bsq swap offers have the deviation value greyed out, but not Fiat offers. This is a live problem already. I just recognized it for the first time.

ripcurlx commented:
Just read the code, that it is done like this to indicate if it is a fixed price or a market price offer. This is not intuitive IMO. I'd rather add the icon as in the offer book and keep the opacity to `1`. @chimp1984 Did you do this to save horizontal space?

ripcurlx committed:
Remove unused import

ripcurlx commented:
> > No idea why the gradle dependencyVerification fails. It is the same content as in the release branch.
> 
> The latest build (re-triggered by @ripcurlx 20 mins ago) was successful. Should also work locally.

This is very weird right now. Sometimes it builds, sometimes it fails because of the verification error.

chimp1984 commented:
@ripcurlx 
Re 1: The BSQ account creation is mentioned in the news popup:
<img width="618" alt="Screenshot 2021-11-03 at 12 03 01" src="https://user-images.githubusercontent.com/54558767/140049361-a5aa689d-52a8-41b3-a28b-94b1d2eb74ab.png">

Re 2:
I cannot reproduce that anymore, but had that issue in the past. It seems some bug with the combobox and with focus. I added a requestFocus call but might be that there are timing/race conditions where the combobox still wins... Maybe some delayed requestFocus can make it more reliable.

Re 3:
Yes, no strong opinion here. But adding a bit explicit naming might help that people understand its different to normal offers/trades. For open offers there is no beavioural difference beside that edit offer is disabled and the disable/enable column which reacts to wallet balance. So that was rather easy to add to the existing one. For trades I think its more justified to add another tab as the process of pending trades->closed trades is not happening for BSQ swaps. 
To add too many tabs might overload a bit the UI and I think its more convenient to have all offers in one place. 

Re 4:
Yes I can rename "BSQ SWAP TRADES".
The trade is basically completed. You cannot do anything to stop it from confirming beside doing a doublespend.
I think we should not give the impression the trade is still in a pending state just because its not confirmed. For BSQ though the funds don't show up before confirmations. So yes that could be improved...
To add the trade after confrimation to history should be possible but comes with some effort. I would prefer to do that in a new PR if considered important.

Re: "I also just recognized that all altcoin/bsq swap offers have the deviation value greyed out, but not Fiat offers. This is a live problem already."
Yes might need a review...

I will add a commit for the rename and add another delayed requestFocus (hope that helps). For the BSQ swap account issue I think it should be covered in the news popup suficiently and maybe we can add/emphasise it in the wiki (have not looked up how its there). If you have a concrete suggestion where to add it additionally let me know.
But would prefer to get that merged and add other changes in new PRs.

chimp1984 committed:
Rename BSQ SWAP TRADES to BSQ Swap Trades

chimp1984 committed:
Add requestFocus call to createOfferView as well
Move method to base class

chimp1984 commented:
Btw. When creating a BSQ altcoin account one get also a popup explaining that there is a default BSQ swaps account...

chimp1984 committed:
Rename tradeAmount to amount in Trade

chimp1984 committed:
Rename getTradePrice to getPrice in Trade

chimp1984 committed:
Rename tradeVolumeProperty to volumeProperty in Trade

chimp1984 committed:
Rename tradeAmountProperty to amountProperty in Trade

chimp1984 committed:
Rename getTradeAmountAsLong to getAmountAsLong in Trade

chimp1984 committed:
Rename setTradePrice to setPriceAsLong in Trade

chimp1984 committed:
Rename getAmount to getAmountAsLong in Trade

chimp1984 committed:
Add getAmountAsLong to TradeModel

chimp1984 committed:
Add getAmount to TradeModel

chimp1984 committed:
Rename getTradeVolume to getVolume in Trade

chimp1984 committed:
Add getVolume to TradeModel

chimp1984 committed:
Add getPrice to TradeModel

chimp1984 committed:
Add default methods for volume, price, amount

chimp1984 committed:
Add getTxFee method to TradeModel

chimp1984 committed:
Rename takerFee to takerFeeAsLong

chimp1984 committed:
Add getTakerFee to TradeModel

chimp1984 committed:
Add getOptionalTxFee method

chimp1984 committed:
Rename makerFee to makerFeeAsLong in BsqSwapTrade

chimp1984 committed:
Add getMakerFee to TradeModel

chimp1984 committed:
Add getOptionalTakerFee and getOptionalMakerFee to Tradeable

chimp1984 committed:
Add null check

chimp1984 committed:
Move ClosedTradableManager and ClosedTradeUtil one level up

We will move BsqSwapTrades into closed trades as well once confrimed

chimp1984 committed:
Refactor methods

Prepare that BsqSwapTrades can be added

chimp1984 committed:
Add info about trade if checkArgument fails

checkArgument(trade instanceof BuyerTrade) failed in dev testing
when clicking very fast confirm buttons.
Likely due lagging UI (previous trade in selection...).
Could not reproduce it afterwards

chimp1984 committed:
Rename CompletedBsqSwapsView* classes to UnconfirmedBsqSwapsView*

chimp1984 committed:
Rename BSQ swap trades tab to Unconfirmed BSQ swaps
Move it left to closed trades

chimp1984 committed:
Rename BSQ Swap Trades to Unconfirmed BSQ Swaps

chimp1984 committed:
Add methods for confirmed and unconfirmed trades

chimp1984 committed:
Use TradeModel instead of Trade in the avatar icon
classes and comparators
Add bsq wallet balance listener to update list
Use BsqSwapTrade details window if Bsq trade
Show "N/A" for deposits in case of BSQ swaps
Cleanup ClosedTradesDataModel

chimp1984 committed:
Add confirmed state
Fix num trades

chimp1984 committed:
Fix getNumPastTrades method

chimp1984 committed:
Rename txFeeAsLong to tradeTxFee and txFeeAsLong to txFeeAsLong
Implement getTxFee in concrete classes. For Maker trades we return the txFee from the offer.

We use txFee in 2 different context:
1. For the txFee the taker pays and which is used for all the trade txs
2. As generic tx fee the user has paid. The makers tx fee is in the offer.
   The taker pays 3 times the tradeTxFee

chimp1984 committed:
Moved AltcoinValidator, FiatPriceValidator and NumberValidator
to bisq.core.util.validation package. No code change.

Is needed for next refactorings.

chimp1984 committed:
Moved PriceUtil to bisq.core.util

chimp1984 committed:
Extract formatter methods to ClosedTradeFormatter

chimp1984 committed:
Moved instance methods to ClosedTradableManager
Renamed ClosedTradeUtil to ClosedTradableUtil

Leaves only static functional methods in ClosedTradableUtil
Domain methods requiring services are in ClosedTradableManager
This keeps also the TradableList

All formatting code is in ClosedTradableFormatter

chimp1984 committed:
Add notification when BSQ swap trade got confirmed

chimp1984 committed:
Fix layout

chimp1984 committed:
Remove ClosedTradableListItem

Was empty wrapper for Tradable

ripcurlx committed:
Fix mixed usage of static import to pacify Codacy

ripcurlx commented:
NIT: If there is no BSQ swap I think we should not show the `Unconfirmed...` tab.

![Bildschirmfoto 2021-11-04 um 09 56 44](https://user-images.githubusercontent.com/170962/140285656-3a35b50f-3e87-4701-a68d-c61dc7e91116.png)

ripcurlx reviewed (approved):
ACK - Besides the NIT I mentioned above I think it is good to move to some extended developer testing on master. Everyone running on master please be aware that the offer propagation might not be very stable until our seednodes have been updated (hint to @Emzy). Besides that BSQ swaps were thoroughly manually and automated tested on Regtest and on Mainnet.
